File: scripts/heygen-generate.mjs
// Generate one avatar video per question and attach it to that question.
//
// Reads every active question with no avatar_url, asks HeyGen to render the
// avatar speaking it, uploads the result to S3, and writes the URL back.
//
// Idempotent: questions that already have a video are skipped, so adding a
// question later renders only that one.
//
//   node scripts/heygen-generate.mjs --dry-run     list what would render
//   node scripts/heygen-generate.mjs --limit 1     render just the first
//   node scripts/heygen-generate.mjs               render everything pending
//
// Needs in .env.local:
//   HEYGEN_API_KEY, HEYGEN_AVATAR_ID, HEYGEN_VOICE_ID
//   HEYGEN_CHARACTER_TYPE   'talking_photo' (default) or 'avatar'
//   DATABASE_URL, AWS_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY,
//   S3_BUCKET_NAME
//
// Written in Node rather than Python so it reuses the dependencies already in
// package.json instead of needing psycopg and boto3.
//
// Uses the HeyGen v3 API with the Avatar IV engine. v2 is deprecated (sunsets
// 2026-10-31) and, more importantly, gave a static talking photo: v3 exposes
// `expressiveness` (which defaults to low, hence the lifeless first cut) and
// `motion_prompt` for body movement and hand gestures.

import { readFileSync } from 'node:fs'
import { neon } from '@neondatabase/serverless'
import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3'

// ── config ───────────────────────────────────────────────────────────────────

function loadEnv() {
  const env = {}
  for (const line of readFileSync('.env.local', 'utf8').split('\n')) {
    const t = line.trim()
    if (!t || t.startsWith('#')) continue
    const i = t.indexOf('=')
    if (i > 0) env[t.slice(0, i).trim()] = t.slice(i + 1).trim()
  }
  return env
}

const env = loadEnv()
const need = key => {
  const v = process.env[key] || env[key]
  if (!v || v.startsWith('YOUR_') || v.startsWith('NEEDS_')) {
    console.error(`${key} is not set in .env.local`)
    process.exit(1)
  }
  return v
}

const HEYGEN_KEY = need('HEYGEN_API_KEY')
const CHARACTER_ID = need('HEYGEN_AVATAR_ID')
const VOICE_ID = need('HEYGEN_VOICE_ID')
const CHARACTER_TYPE = (process.env.HEYGEN_CHARACTER_TYPE || env.HEYGEN_CHARACTER_TYPE || 'talking_photo').trim()
// en-IN makes a multilingual voice speak English with an Indian accent. Without
// it the same voice reads the script in a Western accent.
const LOCALE = (process.env.HEYGEN_LOCALE || env.HEYGEN_LOCALE || '').trim()

// Avatar IV motion controls. expressiveness defaults to 'low' at the API,
// which is what made the avatar look like a still photo with a moving mouth.
const EXPRESSIVENESS = (process.env.HEYGEN_EXPRESSIVENESS || env.HEYGEN_EXPRESSIVENESS || 'high').trim()
const MOTION_PROMPT = (process.env.HEYGEN_MOTION_PROMPT || env.HEYGEN_MOTION_PROMPT ||
  'Speaking warmly and directly to the camera as a friendly college counsellor. ' +
  'Natural hand gestures while explaining, relaxed shoulders, occasional small ' +
  'head movements and nods, engaged and encouraging expression.').trim()
const RESOLUTION = (process.env.HEYGEN_RESOLUTION || env.HEYGEN_RESOLUTION || '1080p').trim()

const DATABASE_URL = need('DATABASE_URL')
const AWS_REGION = need('AWS_REGION')
const S3_BUCKET = need('S3_BUCKET_NAME')

const HEYGEN = 'https://api.heygen.com'
const HEADERS = { 'X-Api-Key': HEYGEN_KEY, 'Content-Type': 'application/json' }

const POLL_TIMEOUT_MS = 15 * 60 * 1000
const POLL_INTERVAL_MS = 10 * 1000

const s3 = new S3Client({
  region: AWS_REGION,
  credentials: {
    accessKeyId: need('AWS_ACCESS_KEY_ID'),
    secretAccessKey: need('AWS_SECRET_ACCESS_KEY'),
  },
})

const sleep = ms => new Promise(r => setTimeout(r, ms))

// ── HeyGen ───────────────────────────────────────────────────────────────────

/**
 * v3 create. Avatar IV animates a photo avatar by look id, so the talking-photo
 * distinction that v2 needed no longer applies.
 */
async function createVideo(text) {
  const res = await fetch(`${HEYGEN}/v3/videos`, {
    method: 'POST',
    headers: HEADERS,
    body: JSON.stringify({
      type: 'avatar',
      avatar_id: CHARACTER_ID,
      script: text,
      voice_id: VOICE_ID,
      resolution: RESOLUTION,
      engine: { type: 'avatar_iv' },
      expressiveness: EXPRESSIVENESS,
      motion_prompt: MOTION_PROMPT,
      ...(LOCALE ? { voice_settings: { locale: LOCALE } } : {}),
    }),
  })
  const body = await res.json().catch(() => ({}))
  if (!res.ok || body.error) {
    throw new Error(`generate failed (${res.status}): ${JSON.stringify(body.error ?? body).slice(0, 400)}`)
  }
  const id = body?.data?.video_id ?? body?.video_id
  if (!id) throw new Error(`no video_id in response: ${JSON.stringify(body).slice(0, 250)}`)
  return id
}

async function waitForVideo(videoId) {
  const deadline = Date.now() + POLL_TIMEOUT_MS
  while (Date.now() < deadline) {
    const res = await fetch(`${HEYGEN}/v3/videos/${videoId}`, { headers: HEADERS })
    const body = await res.json().catch(() => ({}))
    const d = body?.data ?? body ?? {}
    const status = d.status
    if (status === 'completed' || status === 'success') {
      const url = d.video_url ?? d.url ?? d.output?.video_url
      if (!url) throw new Error(`completed but no url: ${JSON.stringify(d).slice(0, 250)}`)
      return url
    }
    if (status === 'failed' || status === 'error') {
      throw new Error(`render failed: ${JSON.stringify(d.error ?? d).slice(0, 300)}`)
    }
    await sleep(POLL_INTERVAL_MS)
  }
  throw new Error('render timed out after 15 minutes')
}

// ── S3 ───────────────────────────────────────────────────────────────────────

/** Private upload. Playback uses presigned URLs, never a public ACL. */
async function uploadToS3(key, bytes) {
  await s3.send(new PutObjectCommand({
    Bucket: S3_BUCKET, Key: key, Body: bytes, ContentType: 'video/mp4',
  }))
  return `https://${S3_BUCKET}.s3.${AWS_REGION}.amazonaws.com/${key}`
}

// ── main ─────────────────────────────────────────────────────────────────────

const dryRun = process.argv.includes('--dry-run')
const limitArg = process.argv.indexOf('--limit')
const limit = limitArg > -1 ? parseInt(process.argv[limitArg + 1]) || 0 : 0
// --bank idle  targets one bank; without it --limit picks by sort_order and can
// grab the wrong row entirely.
const bankArg = process.argv.indexOf('--bank')
const bankFilter = bankArg > -1 ? process.argv[bankArg + 1] : null

// HTTP driver, not the WebSocket Client: renders take minutes each and a
// persistent socket dies mid-run, taking the whole batch down with it.
const sql = neon(DATABASE_URL)

try {
  const pending = bankFilter
    ? await sql`
        select id, bank, position_group, sort_order, content
        from questions
        where avatar_url is null and active and bank = ${bankFilter}
        order by sort_order asc`
    : await sql`
        select id, bank, position_group, sort_order, content
        from questions
        where avatar_url is null and active
        order by sort_order asc`

  const queue = limit ? pending.slice(0, limit) : pending

  if (queue.length === 0) {
    console.log('Nothing to do. Every active question already has a video.')
  } else {
    console.log(`engine:    avatar_iv  expressiveness=${EXPRESSIVENESS}  ${RESOLUTION}`)
    console.log(`avatar:    ${CHARACTER_ID}`)
    console.log(`voice:     ${VOICE_ID}${LOCALE ? `  locale=${LOCALE}` : '  (no locale set)'}\n`)
    console.log(`${queue.length} to render:\n`)
    for (const q of queue) {
      const tag = q.bank === 'behavioral' ? `Q${q.sort_order}` : q.bank.toUpperCase()
      console.log(`  ${tag.padEnd(8)} ${q.content.slice(0, 66).replace(/\n/g, ' ')}…`)
    }

    if (dryRun) {
      console.log('\nDry run. Nothing rendered.')
    } else {
      console.log()
      let failures = 0

      for (const [i, q] of queue.entries()) {
        const tag = q.bank === 'behavioral' ? `Q${q.sort_order}` : q.bank.toUpperCase()
        process.stdout.write(`[${i + 1}/${queue.length}] ${tag} … `)
        try {
          const videoId = await createVideo(q.content)
          process.stdout.write(`rendering(${videoId.slice(0, 8)}) … `)
          const url = await waitForVideo(videoId)

          const bytes = Buffer.from(await (await fetch(url)).arrayBuffer())
          const s3Url = await uploadToS3(`avatars/${q.id}.mp4`, bytes)
          await sql`update questions set avatar_url = ${s3Url} where id = ${q.id}`

          console.log(`done (${(bytes.length / 1e6).toFixed(1)} MB)`)
        } catch (err) {
          failures++
          console.log(`FAILED: ${err.message}`)
        }
      }

      console.log(`\n${queue.length - failures} rendered, ${failures} failed.`)
      if (failures) console.log('Re-run to retry. Successful ones are skipped.')
    }
  }
} catch (err) {
  console.error('ERROR:', err.message)
  process.exitCode = 1
}

File: scripts/heygen_generate.py
"""
Generate one Hinglish avatar video per question and attach it to that question.

Reads every active question with no avatar_url, asks HeyGen to render the
avatar speaking it, uploads the result to S3, and writes the URL back.

Idempotent: questions that already have a video are skipped, so adding a new
question later regenerates only that one.

Run:  python3 scripts/heygen_generate.py
Dry:  python3 scripts/heygen_generate.py --dry-run

Requires in .env.local:
  HEYGEN_API_KEY, HEYGEN_AVATAR_ID, HEYGEN_VOICE_ID
  AWS_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, S3_BUCKET_NAME
  DATABASE_URL  (the Neon connection string)

NOTE: the HeyGen request/response shapes below follow their documented v2 API.
Verify against your account (or the HeyGen MCP server) before the first real
run — vendors move these around, and a wrong field name here just 400s.
"""

import argparse
import pathlib
import sys
import tempfile
import time

import boto3
import psycopg
import requests

# ── Config ────────────────────────────────────────────────────────────────────

ENV_PATH = pathlib.Path(__file__).parent.parent / ".env.local"


def load_env() -> dict:
    if not ENV_PATH.exists():
        sys.exit(f"Missing {ENV_PATH}. Copy .env.local.example and fill it in.")
    env = {}
    for line in ENV_PATH.read_text().splitlines():
        line = line.strip()
        if line and not line.startswith("#") and "=" in line:
            k, v = line.split("=", 1)
            env[k.strip()] = v.strip()
    return env


ENV = load_env()


def need(key: str) -> str:
    val = ENV.get(key)
    if not val or val.startswith("YOUR_"):
        sys.exit(f"{key} is not set in .env.local")
    return val


HEYGEN_KEY = need("HEYGEN_API_KEY")
AVATAR_ID = need("HEYGEN_AVATAR_ID")
VOICE_ID = need("HEYGEN_VOICE_ID")

AWS_REGION = need("AWS_REGION")
AWS_KEY = need("AWS_ACCESS_KEY_ID")
AWS_SECRET = need("AWS_SECRET_ACCESS_KEY")
S3_BUCKET = need("S3_BUCKET_NAME")

DATABASE_URL = need("DATABASE_URL")

HEYGEN_BASE = "https://api.heygen.com"
HEYGEN_HEADERS = {"X-Api-Key": HEYGEN_KEY, "Content-Type": "application/json"}

POLL_TIMEOUT_SEC = 900
POLL_INTERVAL_SEC = 10


# ── Database (Neon) ──────────────────────────────────────────────────────────

def fetch_pending_questions() -> list[dict]:
    """Active questions that have no avatar video yet."""
    with psycopg.connect(DATABASE_URL) as conn:
        rows = conn.execute(
            """
            select id::text, content, position_group, sort_order
            from questions
            where avatar_url is null and active
            order by sort_order asc
            """
        ).fetchall()
    return [
        {"id": r[0], "content": r[1], "position_group": r[2], "sort_order": r[3]}
        for r in rows
    ]


def set_avatar_url(question_id: str, url: str) -> None:
    with psycopg.connect(DATABASE_URL) as conn:
        conn.execute(
            "update questions set avatar_url = %s where id = %s",
            (url, question_id),
        )
        conn.commit()


# ── HeyGen ────────────────────────────────────────────────────────────────────

def create_video(text: str) -> str:
    """Kick off a render. Returns the HeyGen video id."""
    payload = {
        "video_inputs": [
            {
                "character": {
                    "type": "avatar",
                    "avatar_id": AVATAR_ID,
                    "avatar_style": "normal",
                },
                "voice": {
                    "type": "text",
                    "input_text": text,
                    "voice_id": VOICE_ID,
                },
            }
        ],
        "dimension": {"width": 1280, "height": 720},
    }
    r = requests.post(
        f"{HEYGEN_BASE}/v2/video/generate",
        headers=HEYGEN_HEADERS,
        json=payload,
        timeout=60,
    )
    if r.status_code >= 400:
        sys.exit(f"HeyGen generate failed ({r.status_code}): {r.text}")
    return r.json()["data"]["video_id"]


def wait_for_video(video_id: str) -> str:
    """Poll until the render completes. Returns the downloadable URL."""
    deadline = time.time() + POLL_TIMEOUT_SEC
    while time.time() < deadline:
        r = requests.get(
            f"{HEYGEN_BASE}/v1/video_status.get",
            headers=HEYGEN_HEADERS,
            params={"video_id": video_id},
            timeout=30,
        )
        r.raise_for_status()
        data = r.json()["data"]
        status = data.get("status")

        if status == "completed":
            return data["video_url"]
        if status == "failed":
            raise RuntimeError(f"HeyGen render failed: {data.get('error')}")

        time.sleep(POLL_INTERVAL_SEC)

    raise TimeoutError(f"HeyGen render timed out after {POLL_TIMEOUT_SEC}s")


# ── S3 ────────────────────────────────────────────────────────────────────────

s3 = boto3.client(
    "s3",
    region_name=AWS_REGION,
    aws_access_key_id=AWS_KEY,
    aws_secret_access_key=AWS_SECRET,
)


def upload(local_path: pathlib.Path, key: str) -> str:
    """Upload privately. Playback goes through presigned URLs, not public ACLs."""
    s3.upload_file(
        str(local_path),
        S3_BUCKET,
        key,
        ExtraArgs={"ContentType": "video/mp4"},
    )
    return f"https://{S3_BUCKET}.s3.{AWS_REGION}.amazonaws.com/{key}"


# ── Main ──────────────────────────────────────────────────────────────────────

def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--dry-run", action="store_true",
                        help="List what would be generated, then stop.")
    parser.add_argument("--limit", type=int, default=0,
                        help="Only process the first N questions.")
    args = parser.parse_args()

    questions = fetch_pending_questions()
    if args.limit:
        questions = questions[: args.limit]

    if not questions:
        print("Nothing to do — every active question already has a video.")
        return

    print(f"{len(questions)} question(s) need an avatar video:\n")
    for q in questions:
        preview = q["content"][:70].replace("\n", " ")
        print(f"  [{q['position_group']}] {preview}…")

    if args.dry_run:
        print("\nDry run — nothing generated.")
        return

    print()
    failures = 0

    for i, q in enumerate(questions, 1):
        qid = q["id"]
        print(f"[{i}/{len(questions)}] {qid}")

        try:
            video_id = create_video(q["content"])
            print(f"    rendering ({video_id})…")
            download_url = wait_for_video(video_id)

            with tempfile.NamedTemporaryFile(suffix=".mp4", delete=False) as tmp:
                resp = requests.get(download_url, stream=True, timeout=300)
                resp.raise_for_status()
                for chunk in resp.iter_content(1 << 20):
                    tmp.write(chunk)
                tmp_path = pathlib.Path(tmp.name)

            s3_url = upload(tmp_path, f"avatars/{qid}.mp4")
            tmp_path.unlink(missing_ok=True)

            set_avatar_url(qid, s3_url)
            print(f"    done → {s3_url}")

        except Exception as exc:  # keep going; one bad question shouldn't stop the batch
            failures += 1
            print(f"    FAILED: {exc}")

    done = len(questions) - failures
    print(f"\n{done} generated, {failures} failed.")
    if failures:
        print("Re-run to retry the failures — successful ones are skipped.")


if __name__ == "__main__":
    main()
